### PR TITLE
changes the installer-pod to write optional pods

### PR DESCRIPTION
**Pull request by p0lyn0mial**
up until now, the installer pod expected only a single pod manifest in the configmap (the cm is specified by the --pod flag)

since we are planning to add a startup monitor for assessing operand's health check we are extending the installer
to take into account optional keys that might contain other pods.

the new pods will be written to the same locations as the main pod (revisioned, manifests).
the name will be derived from the name filed and suffixed with "-pod.yaml"

xref: https://github.com/openshift/library-go/pull/1130

### CODE REVIEW

**Comment by p0lyn0mial** [comment]
untested

/assign @sttts

**Comment by p0lyn0mial** [comment]
proof PR https://github.com/openshift/cluster-kube-apiserver-operator/pull/1175

File: pkg/operator/staticpod/installerpod/cmd.go [hunk 1]
```diff
@@ -300,55 +300,37 @@ func (o *InstallOptions) copyContent(ctx context.Context) error {
...
-		return err
-	}
-	klog.Infof("Writing static pod manifest %q ...\n%s", path.Join(o.PodManifestDir, podFileName), finalPodBytes)
-	if err := ioutil.WriteFile(path.Join(o.PodManifestDir, podFileName), []byte(finalPodBytes), 0644); err != nil {
-		return err
-	}
+	// then write the required pod and all optional
+	// the key must contain pod.yaml to be considered
+	for rawPodKey, rawPod := range podsConfigMap.Data {
+		if !strings.Contains(rawPodKey, "pod.yaml") {
```
Comment (sttts) [inline pkg/operator/staticpod/installerpod/cmd.go:319]: What else do you expect here?

File: pkg/operator/staticpod/installerpod/cmd.go [hunk 2]
```diff
@@ -300,55 +300,37 @@ func (o *InstallOptions) copyContent(ctx context.Context) error {
...
+			continue
+		}
+		var podName string
+		if rawPodKey == "pod.yaml" {
+			// this is our main pod, assign the previous name for compatibility
+			podName = o.PodConfigMapNamePrefix + ".yaml"
+		}
```
Comment (sttts) [inline pkg/operator/staticpod/installerpod/cmd.go:326]: can't we have a fully qualified name as key? e.g. kube-apiserver-pod.yaml.

File: pkg/operator/staticpod/installerpod/cmd.go [hunk 3]
```diff
@@ -300,55 +300,37 @@ func (o *InstallOptions) copyContent(ctx context.Context) error {
...
-	if err := ioutil.WriteFile(path.Join(o.PodManifestDir, podFileName), []byte(finalPodBytes), 0644); err != nil {
-		return err
-	}
+	// then write the required pod and all optional
+	// the key must contain pod.yaml to be considered
+	for rawPodKey, rawPod := range podsConfigMap.Data {
+		if !strings.Contains(rawPodKey, "pod.yaml") {
+			continue
+		}
+		var podName string
```
Comment (sttts) [inline pkg/operator/staticpod/installerpod/cmd.go:322]: this is not the podName, but the manifestFileName

File: pkg/operator/staticpod/installerpod/cmd.go [hunk 4]
```diff
@@ -300,55 +300,37 @@ func (o *InstallOptions) copyContent(ctx context.Context) error {
...
+		if rawPodKey == "pod.yaml" {
+			// this is our main pod, assign the previous name for compatibility
+			podName = o.PodConfigMapNamePrefix + ".yaml"
+		}
 
+		klog.Infof("Writing a pod under %q key \n%s", rawPodKey, rawPod)
+		err := o.writePod([]byte(rawPod), podName, resourceDir)
```
Comment (sttts) [inline pkg/operator/staticpod/installerpod/cmd.go:329]: nit: dir, manifestFileName

File: pkg/operator/staticpod/installerpod/cmd.go [hunk 5]
```diff
@@ -300,55 +300,37 @@ func (o *InstallOptions) copyContent(ctx context.Context) error {
...
-	} else if !os.IsNotExist(err) {
-		return err
-	}
-	klog.Infof("Writing static pod manifest %q ...\n%s", path.Join(o.PodManifestDir, podFileName), finalPodBytes)
-	if err := ioutil.WriteFile(path.Join(o.PodManifestDir, podFileName), []byte(finalPodBytes), 0644); err != nil {
-		return err
-	}
+	// then write the required pod and all optional
+	// the key must contain pod.yaml to be considered
+	for rawPodKey, rawPod := range podsConfigMap.Data {
```
Comment (sttts) [inline pkg/operator/staticpod/installerpod/cmd.go:318]: does order matter?

Comment (p0lyn0mial) [reply pkg/operator/staticpod/installerpod/cmd.go:318]: I was thinking about the same, I can make it stable.
At the moment it doesn't matter, the `startup-monitor` will wait until an operand matches its current revision.

File: pkg/operator/staticpod/installerpod/cmd.go (hunk 3 above)
Comment (p0lyn0mial) [reply pkg/operator/staticpod/installerpod/cmd.go:322]: true :)

File: pkg/operator/staticpod/installerpod/cmd.go (hunk 2 above)
Comment (p0lyn0mial) [reply pkg/operator/staticpod/installerpod/cmd.go:326]: oh, sure, we could, good idea, will change.

File: pkg/operator/staticpod/installerpod/cmd.go (hunk 1 above)
Comment (p0lyn0mial) [reply pkg/operator/staticpod/installerpod/cmd.go:319]: as of today the `cm` contains a few additional keys, like `forceRedeploymentReason`

File: pkg/operator/staticpod/installerpod/cmd.go (hunk 2 above)
Comment (p0lyn0mial) [reply pkg/operator/staticpod/installerpod/cmd.go:326]: I think that this is a two-step process, first, we need to merge this PR, and then we need to update `kas-o`, right ?

File: pkg/operator/staticpod/installerpod/cmd.go (hunk 5 above)
Comment (sttts) [reply pkg/operator/staticpod/installerpod/cmd.go:318]: I also think kubelet is not deterministic in any way. So I think we are fine.

File: pkg/operator/staticpod/installerpod/cmd.go (hunk 1 above)
Comment (sttts) [reply pkg/operator/staticpod/installerpod/cmd.go:319]: would check for suffix `-pod.yaml`.

File: pkg/operator/staticpod/installerpod/cmd.go (hunk 2 above)
Comment (sttts) [reply pkg/operator/staticpod/installerpod/cmd.go:326]: probably, yes.

Comment (p0lyn0mial) [reply pkg/operator/staticpod/installerpod/cmd.go:326]: ok, I will prepare a follow up PRs

**Pull request merged**